### DISCUSSION TITLE
Basic git question

### DISCUSSION THREAD

**Discussion by isaacdvory** [community/community#22794] [New to GitHub]
<p>Hi,<br>
I’m very much a noob and I’m reading through the manual but it’s not super clear to me,  so I’ll appreciate the extra help :slightly_smiling_face:   (currently on this stage: <a href="https://git-scm.com/book/en/v2/Git-Basics-Recording-Changes-to-the-Repository" rel="nofollow noopener">https://git-scm.com/book/en/v2/Git-Basics-Recording-Changes-to-the-Repository</a>)</p>
<p>I never had git , just got started, to my understanding the first thing is to create the repo,<br>
so I’m currently on my LIVE server, with my LIVE files, and I did - git init, added all files I wanted to gitignore, then did git add * and git commit.</p>
<ol>
<li>Did I just create repo? is it just a local repo at the moment or it’s not called a repo yet?</li>
</ol>
<p>After I did the first commit,  I realized that some of the files had sensitive information in them , I still need these files to get tracked, so I edited them , remove the sensitive info to a different untracked file , did git add again and git commit</p>
<ol start="2">
<li>how do I remove the first commit ? I don’t want the original one with the sensitive info in the files.</li>
<li>if I do git remove, will it remove files from my LIVE directory?</li>
<li>assuming I did everything correct until now, and that I can fix what I asked here.<br>
I understand the next step would be to upload it somewhere (push?) , then pull it in my STAGING server, and from that point all the changes I make, all the commits, and push should be done just from there and the only thing I’m suppose to do on my live server is PULL?</li>
</ol>
<p>Thanks!</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="isaacdvory" data-post="1" data-topic="130274">
<div class="title">
<div class="quote-controls"></div>
 isaacdvory:</div>
<blockquote>
<p>did git add * and git commit.</p>
</blockquote>
</aside>
<p><code>git add *</code> works, but is generally bad practice, because it’s easy to accidentally commit files with sensitive information, or that otherwise don’t belong into git, like it happened to you. I highly recommend adding files one by one or in small sets, and checking what you added with <code>git diff --cached</code> before committing.</p>
<aside class="quote no-group" data-username="isaacdvory" data-post="1" data-topic="130274">
<div class="title">
<div class="quote-controls"></div>
 isaacdvory:</div>
<blockquote>
<ol>
<li>Did I just create repo? is it just a local repo at the moment or it’s not called a repo yet?</li>
</ol>
</blockquote>
</aside>
<p>It’s a local repository, yes.</p>
<aside class="quote no-group" data-username="isaacdvory" data-post="1" data-topic="130274">
<div class="title">
<div class="quote-controls"></div>
 isaacdvory:</div>
<blockquote>
<p>After I did the first commit, I realized that some of the files had sensitive information in them , I still need these files to get tracked, so I edited them , remove the sensitive info to a different untracked file , did git add again and git commit</p>
<ol start="2">
<li>how do I remove the first commit ? I don’t want the original one with the sensitive info in the files.</li>
</ol>
</blockquote>
</aside>
<p>In general: You can change the current HEAD commit by <code>add</code>ing the changes you want, and running <code>git commit --amend</code>. The tricky think is that you’ll need <code>git rebase</code> to go back in the history and make changes there (but the rest works the same). Considering you don’t have any history you want to keep, you can just delete the <code>.git/</code> directory and start fresh with <code>git init</code> (that second <code>git init</code> call detected you already had a repository and thus didn’t do anything).</p>
<p>However, I strongly recommend you do <strong>not</strong> do that on a live server, even if you were a highly experienced git user. What I suggest you do is copy the files to another system and create the repository there. That way any experiments or mistakes won’t break your live server.</p>
<aside class="quote no-group" data-username="isaacdvory" data-post="1" data-topic="130274">
<div class="title">
<div class="quote-controls"></div>
 isaacdvory:</div>
<blockquote>
<ol start="3">
<li>if I do git remove, will it remove files from my LIVE directory?</li>
</ol>
</blockquote>
</aside>
<p>By default, yes. You can use <code>git rm --cached</code> to tell git to remove a file in the next commit without touching the working tree (the directory you’re working with). But that will not remove the file from history.</p>
<aside class="quote no-group" data-username="isaacdvory" data-post="1" data-topic="130274">
<div class="title">
<div class="quote-controls"></div>
 isaacdvory:</div>
<blockquote>
<ol start="4">
<li>assuming I did everything correct until now, and that I can fix what I asked here.<br>
I understand the next step would be to upload it somewhere (push?) , then pull it in my STAGING server, and from that point all the changes I make, all the commits, and push should be done just from there and the only thing I’m suppose to do on my live server is PULL?</li>
</ol>
</blockquote>
</aside>
<p>You can do that, assuming there are no more history modifications (if there are, a simple <code>pull</code> might not work). I want to emphasize though: There is no <em>technical</em> reason to only push from one place, in fact git is specifically made to allow many people to edit in many places and merge the results. In terms of <em>organization</em> it’s generally a good thing not to edit things on live servers.</p>
<p>However, you might want to consider setting up a deployment step, that only copies the files that are needed onto your live server instead of having the whole repository there. But what’s best there really depends on what your specific needs are. :slightly_smiling_face:</p>

**Comment by isaacdvory** [comment]
<p>Thank you so much for your answer.<br>
What I’m trying to do with my initial git setup is to have my first repo, I did that with git init, git add * (because I have thousands of files, but I did fill out .gitignore first)<br>
all on the live files, because that’s all I have, I’m doing all of this so I can then setup a staging area on a different server.<br>
From reading your answer I understand that what I should actually do is move the files Manually to the new server and start my git from there, not from the live files?<br>
If so, I can just delete the .git folder as you said, move files manually and restart the process on on the new server?</p>
<p>then on the live server I would need to just</p>
<ol>
<li>connect a remote repo</li>
<li>normally only do Pull/fetch on that live server?<br>
I’m not sure I understand the last paragraph on your answer with the deployment step</li>
</ol>
<p>thanks.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="isaacdvory" data-post="3" data-topic="130274">
<div class="title">
<div class="quote-controls"></div>
 isaacdvory:</div>
<blockquote>
<p>What I’m trying to do with my initial git setup is to have my first repo, I did that with git init, git add * (because I have thousands of files, but I did fill out .gitignore first)</p>
</blockquote>
</aside>
<p>Ah yes, in that case it’ll be difficult to manually check everything. Though it also makes me wonder if all those files are part of the same project, or if you might want to split them into smaller sets. But I don’t know your situation. :slightly_smiling_face:</p>
<aside class="quote no-group" data-username="isaacdvory" data-post="3" data-topic="130274">
<div class="title">
<div class="quote-controls"></div>
 isaacdvory:</div>
<blockquote>
<p>From reading your answer I understand that what I should actually do is move the files Manually to the new server and start my git from there, not from the live files?</p>
</blockquote>
</aside>
<p>Yes, that’s what I recommend, because that way any experiments or mistakes while you figure things out won’t affect the live server (which I assume you want to keep running without interruptions). For git it doesn’t matter either way, git just tracks the files, wherever you tell it to. :wink:</p>
<aside class="quote no-group" data-username="isaacdvory" data-post="3" data-topic="130274">
<div class="title">
<div class="quote-controls"></div>
 isaacdvory:</div>
<blockquote>
<p>If so, I can just delete the .git folder as you said, move files manually and restart the process on on the new server?</p>
</blockquote>
</aside>
<p>Yes. Keep in mind that this will delete the complete history of the repository, so I usually wouldn’t recommend it. But you said you had just one commit that you explicitly want to get rid of and the current fixed state, so in this case it seems appropriate to simply start over.</p>
<aside class="quote no-group" data-username="isaacdvory" data-post="3" data-topic="130274">
<div class="title">
<div class="quote-controls"></div>
 isaacdvory:</div>
<blockquote>
<ol>
<li>connect a remote repo</li>
<li>normally only do Pull/fetch on that live server?</li>
</ol>
</blockquote>
</aside>
<p>That would work, yes.</p>
<aside class="quote no-group" data-username="isaacdvory" data-post="3" data-topic="130274">
<div class="title">
<div class="quote-controls"></div>
 isaacdvory:</div>
<blockquote>
<p>I’m not sure I understand the last paragraph on your answer with the deployment step</p>
</blockquote>
</aside>
<p>“Deployment” is a generic term, in IT it refers to making something (like a software of website) operational. Pulling the latest version from git on your live server would be one possible way to <em>deploy</em> your website, if the server just needs the files from git as they are.</p>
<p>What I’m questioning is whether that’s the best way to do it. Again, I don’t know your system, so I can’t say for sure. But for example you could alternatively create a TAR or ZIP archive of all the files in your repository using <code>git archive</code>, copy that to the live server, and unpack. That way the server doesn’t need to store the whole git history. If you need to compile anything it’s usually better to do that outside the live server, test, and deploy (which would then include copying the binaries) only after successful test. Automating all of that is commonly referred to as Continous Integration (for automatic building and tests), and Continous Deployment (for automated deployment after), but that’s a much bigger topic.</p>
<p>I recently read a report about companies that <em>unintentionally</em> published the whole git history with their websites, and in some cases those histories contained secrets, so I had to think of that here. :wink:</p>

**Comment by isaacdvory** [comment]
<p>Thanks again.<br>
It’s a Django project , essentially one big site, I think it’s OK for it to be under one project.<br>
as for the deployment, again as it’s a django project, the files are just as they are , no compile is needed, my very basic idea is to test on my staging server and when I comfortable it works as intended, pull to live.<br>
The site is for internal company use only, and the repo I intend to keep locally on the staging server only and using it as the remote as well.</p>

**Reply by Lokita233** [reply]
Hajer

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="isaacdvory" data-post="5" data-topic="130274">
<div class="title">
<div class="quote-controls"></div>
 isaacdvory:</div>
<blockquote>
<p>The site is for internal company use only, and the repo I intend to keep locally on the staging server only and using it as the remote as well.</p>
</blockquote>
</aside>
<p>In that case a push/pull workflow should work just fine. Just make sure your web server is not going to serve the <code>.git/</code> directory to keep it internal. :slightly_smiling_face:</p>

**Reply by Lokita233** [reply]
@Lokita233

**Comment by isaacdvory** [comment]
<p>Hi :slight_smile: it’s been a while and I need some help again please.<br>
I’ve finished with the staging server , I have my site in folder:</p>
<p>/home/dev_staging/staging/<br>
and git directory in /home/dev_staging/staging/.git</p>
<p>I made sure it’s not being served by nginx.<br>
I’ve also added a user “git” to the linux machine, I added a new group gitstating<br>
added that user to the group, and gave the group access to my staging files.<br>
Next I added my public keys to the git user …</p>
<p>Went back to my live server, I have a staging folder there as well in  /home/deployment/site/<br>
in that folder I did git init, which create .git folder, then I added the remote with git remove add staging ssh://git@STAGINGIP:PORT/home/dev_staging/staging</p>
<p>then did “git staging fetch master”, it asks for a password and then :</p>
<ul>
<li>branch            master     -&gt; FETCH_HEAD</li>
</ul>
<p>Seems good right? so why when I do git status , it just shows that all my files are untracked? is that normal? shouldn’t it have recognized that it’s the same files and show my with files are different etc…?</p>
<p>or I should still do in this second staging server the same as I did in the first staging? add .gitignore, do git add , commit ?  I don’t understand how do I start working with this normally as everybody works with git :frowning:</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="isaacdvory" data-post="7" data-topic="130274">
<div class="title">
<div class="quote-controls"></div>
 isaacdvory:</div>
<blockquote>
<p>then did “git staging fetch master”, it asks for a password and then :</p>
<ul>
<li>branch master -&gt; FETCH_HEAD</li>
</ul>
<p>Seems good right? so why when I do git status , it just shows that all my files are untracked? is that normal? shouldn’t it have recognized that it’s the same files and show my with files are different etc…?</p>
</blockquote>
</aside>
<p>That is expected: <code>fetch</code> only updates the information git has about the remote repository, not the local branches, much less files. If I understand correctly you don’t have any local commits, so you could simply <a href="https://git-scm.com/docs/git-reset#Documentation/git-reset.txt-emgitresetemltmodegtltcommitgt"><code>reset</code> your current local branch to the state of <code>staging/master</code></a>:</p>
<pre><code class="lang-bash">git reset staging/master
</code></pre>
<p>After that you should see any differences with the working tree files.</p>
<p>A word of caution: I’d be worried about leaving an SSH key on the live server. It looks like you use it to do a regular login and then access files, so someone who steals it might be able to get other data, too. I assume the live server has a higher attack surface (especially if it’s on the internet) than your work system or the staging server, so I’d rather do login the other way around and push to the live server. :wink:</p>

**Comment by isaacdvory** [comment]
<p>Thanks, I did the reset now, I think it’s fine.<br>
As for the key, the key was added to the staging server , not to the live server, in the live server I login with the username and password of GIT staging user , without a key.<br>
now I will request from my devs to give me their public keys, so I can add them as well to the staging server so they can make changes to the git repo.</p>
<p>thanks again</p>

**Comment by isaacdvory** [comment]
<p>Hi,</p>
<p>I got some error message when I tried to push from the live server,  that I understand is because I didn’t create a bare repo on the repo server, so I had to do this: “git config receive.denyCurrentBranch ignore” in the repo server<br>
then when I did the push it worked, but now on the staging (repo) server, git status is like this:</p>
<p>On branch master<br>
Changes to be committed:<br>
(use “git restore --staged …” to unstage)<br>
modified:   base/models.py<br>
modified:   orders/cost_calculator.py<br>
modified:   orders/forms.py<br>
modified:   orders/tasks.py<br>
modified:   orders/views.py<br>
deleted:    stockassist/service_providers/<strong>init</strong>.py</p>
<p>and I don’t understand what do I need to do now? commit again ?</p>

**Comment by airtower-luna** [comment]
<p>That’s because you pushed changes from the live server to the staging server. Pushing with <code>receive.denyCurrentBranch ignore</code> means only the branch (in your case <code>master</code>) is updated, not the index or working tree. So now the files that you changed in the commit from the live server are different between the <code>master</code> branch and the files you have on the staging server.</p>
<p>As usual what to do depends on which version you actually want to keep. If you want to keep the version of the staging server, you should discard the commit made on the live server. If you want to keep the version from the live server, call <code>git checkout</code> on the changed/missing files on the staging server to get them from the current HEAD commit.</p>
<p>That said, based on what you said about the intended workflow I’m confused why you made a commit on the live server at all. Didn’t you want to use the code exactly as it has been tested in staging?</p>

**Comment by isaacdvory** [comment]
<p>It’s because between the time it took me to move files manually to the staging server and setting up everything to work,  some files already changed in the live server, so after fetching from the staging to live , I saw the files that are newer in the live server, added them, committed and pushed</p>

**Comment by isaacdvory** [comment]
<p>so how do I actually do the git checkout?</p>

**Comment by airtower-luna** [comment, marked as answer]
<p><code>git checkout FILE</code> will overwrite the working tree file with the current HEAD version, so for example:</p>
<pre><code class="lang-bash">git checkout base/models.py
</code></pre>
<p>You can give multiple files at once. Alternatively you could use <code>git reset --hard HEAD</code> to reset <em>all</em> files in the working tree to the HEAD state. I prefer the checkout approach because it does less damage in case of a mistake. :wink:</p>
